lib/asyncSender.ts

```diff
@@ -0,0 +1,181 @@
+          this.connection.id, id, this.name, this.session.id, deleteResult
+        );
+        return promise.resolve(context.delivery);
+      }
+    };
+
+    const onFailure = (eventName: string, id: number, error?: Error) => {
+      if (this.deliveryDispositionMap.has(id)) {
+        log.sender(
+          "[%s] Event: '%s', Found the delivery with id %d in the map of " +
+          "sender '%s' on amqp session '%s' and cleared the timer.",
+          this.connection.id, eventName, id, this.name, this.session.id
+        );
+        const promise = this.deliveryDispositionMap.get(id) as PromiseLike;
+        clearTimeout(promise.timer);
+        const deleteResult = this.deliveryDispositionMap.delete(id);
+        log.sender(
+          "[%s] Event: '%s', Successfully deleted the delivery with id %d from the " +
+          " map of sender '%s' on amqp session '%s' and cleared the timer: %s.",
+          this.connection.id, eventName, id, this.name, this.session.id, deleteResult
+        );
+        const msg = `Sender '${this.name}' on amqp session '${this.session.id}', received a ` +
+          `'${eventName}' disposition. Hence we are rejecting the promise.`;
+        const err = error || new Error(msg);
+        log.sender("[%s] %s", this.connection.id, msg);
+        return promise.reject(err);
+      }
+    };
+
+    const onError = (eventName: string, error?: Error) => {
+      for (const id of this.deliveryDispositionMap.keys()) {
+        onFailure(eventName, id, error);
+      }
+    };
+
+    this.on(SenderEvents.accepted, onSuccess);
+    this.on(SenderEvents.rejected, (context: EventContext) => {
+      onFailure(SenderEvents.rejected, context.delivery!.id, context!.delivery!.remote_state!.error);
+    });
+    this.on(SenderEvents.released, (context: EventContext) => {
+      onFailure(SenderEvents.released, context.delivery!.id, context!.delivery!.remote_state!.error);
+    });
+    this.on(SenderEvents.modified, (context: EventContext) => {
+      onFailure(SenderEvents.modified, context.delivery!.id, context!.delivery!.remote_state!.error);
+    });
+
+    // The user may have it's custom reconnect logic for bringing the sender link back online and
+    // retry logic for sending messages on failures hence they can provide their error handlers
+    // for sender_error and session_error.
+    // If the user did not provide their error handler for sender_error and session_error, then we
+    // add our handlers and make sure we clear the timer and reject the promise for sending
+    // messages with appropriate Error.
+    if (!options.onError) {
+      this.on(SenderEvents.senderError, (context: EventContext) => {
+        onError(SenderEvents.senderError, context.sender!.error as Error);
+      });
+    }
+    if (!options.onSessionError) {
+      this.session.on(SessionEvents.sessionError, (context: EventContext) => {
+        onError(SessionEvents.sessionError, context.sender!.error as Error);
```

Comment:
Shouldn't `context.sender!.error` be `context.session!.error`?

Reply:
yup. Thanks for catching this :)